{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b1da87d2-25a3-41ee-984b-e32be6043c1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as st\n",
    "import vcf\n",
    "import vcf.filters\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import gzip\n",
    "import shutil\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82d8b21e-ff22-4eda-9975-dc54a96e90c9",
   "metadata": {},
   "source": [
    "# Extracting and analyzing data from the VCF files\n",
    "\n",
    "## Don't need to do any QC filtering because that's for reads from FASTQ files\n",
    "\n",
    "## These are all post-QC sequences in FASTA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0f7ed397-b340-4a8a-923c-1004df49954c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_vcf_snps(fName, strain_diff_df, strain_aln, other_strain):\n",
    "    \n",
    "    if \"G_\" in fName:\n",
    "        chrom = \"gene=G\"\n",
    "        protein = \"G\"\n",
    "    elif \"P_\" in fName:\n",
    "        chrom = \"gene=P\"\n",
    "        protein = \"P\"\n",
    "        \n",
    "    strain_diff_protein = strain_diff_df.query(\"Protein==@protein\").reset_index(drop=True)\n",
    "    \n",
    "    if strain_aln not in [\"BGD\", \"MYS\"]:\n",
    "        raise ValueError(f\"{strain_aln} is not a valid strain name!\")\n",
    "    if other_strain not in [\"BGD\", \"MYS\"]:\n",
    "        raise ValueError(f\"{other_strain} is not a valid strain name!\")\n",
    "    \n",
    "    pos = []\n",
    "    ref = []\n",
    "    alt = []\n",
    "    strain_diffs = []\n",
    "    \n",
    "    vcf_reader = vcf.Reader(open(fName, 'r'))\n",
    "\n",
    "    for record in vcf_reader:\n",
    "        \n",
    "        # check that all variants are coming from the correct gene. Might be overkill, but just another sanity check\n",
    "        if chrom not in record.CHROM:\n",
    "            raise ValueError(f\"SNP found outside of {chrom}\")\n",
    "            \n",
    "        # convert form VCF objects to strings to do string comparison\n",
    "        alt_alleles = [str(char) for char in record.ALT]\n",
    "\n",
    "        if \"<*>\" in alt_alleles:\n",
    "            alt_alleles.remove(\"<*>\")\n",
    "        alt_alleles = \"\".join(alt_alleles)\n",
    "            \n",
    "        pos.append(record.POS)\n",
    "        ref.append(record.REF)\n",
    "        alt.append(alt_alleles)\n",
    "        \n",
    "        if record.POS in strain_diff_protein.Position.values:\n",
    "            pos_of_interest = strain_diff_protein.loc[strain_diff_protein.Position==record.POS, :]\n",
    "        \n",
    "            if record.REF == pos_of_interest[strain_aln].values[0] and len(alt_alleles) == 1:\n",
    "                if alt_alleles == pos_of_interest[other_strain].values[0]:\n",
    "                    strain_diffs.append(1)\n",
    "                else:\n",
    "                    strain_diffs.append(0)\n",
    "            else:\n",
    "                strain_diffs.append(0)\n",
    "        else:\n",
    "            strain_diffs.append(0)\n",
    "        \n",
    "    return pd.DataFrame({\"pos\": pos, \"ref\": ref, \"alt\": alt, \"strain_diff\": strain_diffs, \"aln\": strain_aln})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5b65cdf0-2734-460b-86db-6bc50f132d8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_vcf_and_metrics(fName, strain_diff_df, strain_aln, other_strain):\n",
    "    \n",
    "    res = parse_vcf_snps(fName, strain_diff_df, strain_aln, other_strain)\n",
    "    non_strain_diffs = len(res.query(\"strain_diff==0\"))/len(res)\n",
    "    \n",
    "    print(f\"{non_strain_diffs} of SNPs are not strain-level variations\")\n",
    "    \n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "147fb336-4c22-47eb-9407-b6da0f8a0e9c",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'strain_diff' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [4]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m g_bgd \u001b[38;5;241m=\u001b[39m parse_vcf_and_metrics(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124malignments/G_BGD_SNPs.vcf\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[43mstrain_diff\u001b[49m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBGD\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMYS\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      2\u001b[0m g_mys \u001b[38;5;241m=\u001b[39m parse_vcf_and_metrics(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124malignments/G_MYS_SNPs.vcf\u001b[39m\u001b[38;5;124m'\u001b[39m, strain_diff, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMYS\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBGD\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      4\u001b[0m p_bgd \u001b[38;5;241m=\u001b[39m parse_vcf_and_metrics(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124malignments/P_BGD_SNPs.vcf\u001b[39m\u001b[38;5;124m'\u001b[39m, strain_diff, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBGD\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMYS\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'strain_diff' is not defined"
     ]
    }
   ],
   "source": [
    "g_bgd = parse_vcf_and_metrics('alignments/G_BGD_SNPs.vcf', strain_diff, \"BGD\", \"MYS\")\n",
    "g_mys = parse_vcf_and_metrics('alignments/G_MYS_SNPs.vcf', strain_diff, \"MYS\", \"BGD\")\n",
    "\n",
    "p_bgd = parse_vcf_and_metrics('alignments/P_BGD_SNPs.vcf', strain_diff, \"BGD\", \"MYS\")\n",
    "p_mys = parse_vcf_and_metrics('alignments/P_MYS_SNPs.vcf', strain_diff, \"MYS\", \"BGD\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fe2408e-9125-4559-be1d-cfd557e0d52d",
   "metadata": {},
   "source": [
    "### Not sure what the difference in proportion between BGD and MYS means. \n",
    "\n",
    "### It is already ignoring strain-level variations (comparing the 2 reference genomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8034bb5d-2246-4d27-a041-5eeaded4f4f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "glyco_bgd_ref = [seq_record.seq for seq_record in SeqIO.parse(\"sequences/genome/AY988601.1_BGD_CDS.fna\", \"fasta\") if \"gene=G\" in seq_record.id][0]\n",
    "glyco_mys_ref = [seq_record.seq for seq_record in SeqIO.parse(\"sequences/genome/NC_00278.1_MYS_CDS.fna\", \"fasta\") if \"gene=G\" in seq_record.id][0]\n",
    "\n",
    "phospho_bgd_ref = [seq_record.seq for seq_record in SeqIO.parse(\"sequences/genome/AY988601.1_BGD_CDS.fna\", \"fasta\") if \"gene=P\" in seq_record.id][0]\n",
    "phospho_mys_ref = [seq_record.seq for seq_record in SeqIO.parse(\"sequences/genome/NC_00278.1_MYS_CDS.fna\", \"fasta\") if \"gene=P\" in seq_record.id][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "27ffd37c-e1f5-4ff6-94ec-8b2955be8a17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1809 1809\n",
      "2130 2130\n"
     ]
    }
   ],
   "source": [
    "print(len(glyco_bgd_ref), len(glyco_mys_ref))\n",
    "print(len(phospho_bgd_ref), len(phospho_mys_ref))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "280f54aa-2cef-4bee-847e-e1735084bbf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "glyco_snps = pd.concat([g_bgd.query(\"strain_diff==0\"), g_mys.query(\"strain_diff==0\")]).reset_index(drop=True)\n",
    "phospho_snps = pd.concat([p_bgd.query(\"strain_diff==0\"), p_mys.query(\"strain_diff==0\")]).reset_index(drop=True)\n",
    "\n",
    "del glyco_snps[\"strain_diff\"]\n",
    "del phospho_snps[\"strain_diff\"]\n",
    "\n",
    "# lengths of all sequences, sanity check again\n",
    "assert glyco_snps.pos.max() <= 1809 and phospho_snps.pos.max() <= 2130"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "de6034dc-d107-40d8-ad2f-07e5597f9625",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_codon(pos, ref_seq, alt):\n",
    "    \n",
    "    # create the mutated sequence, strings and Seqs are immutable\n",
    "    alt_seq = list(ref_seq)\n",
    "    alt_seq[pos-1] = alt\n",
    "    alt_seq = Seq(\"\".join(alt_seq))\n",
    "    \n",
    "    # position is 1-indexed, make sure to use 0-indexing here\n",
    "    if pos % 3 == 0:\n",
    "        codon = ref_seq[pos-3:pos]\n",
    "        alt_codon = alt_seq[pos-3:pos]\n",
    "    elif pos % 3 == 1:\n",
    "        codon = ref_seq[pos-1:pos+2]\n",
    "        alt_codon = alt_seq[pos-1:pos+2]\n",
    "    else:\n",
    "        codon = ref_seq[pos-2:pos+1]\n",
    "        alt_codon = alt_seq[pos-2:pos+1]\n",
    "        \n",
    "    return codon.translate(), alt_codon.translate()\n",
    "\n",
    "\n",
    "def get_aa_num(pos):\n",
    "    \n",
    "    if pos < 1:\n",
    "        raise ValueError(\"Position must be a natural number!\")\n",
    "    \n",
    "    if pos % 3 == 0:\n",
    "        return int(pos / 3)\n",
    "    else:\n",
    "        return int(pos / 3) + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "88643318-6dc0-4adf-b43d-c1bdea28f216",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, row in glyco_snps.iterrows():\n",
    "    \n",
    "    if row[\"aln\"] == \"BGD\":\n",
    "        ref_name = glyco_bgd_ref\n",
    "    else:\n",
    "        ref_name = glyco_mys_ref\n",
    "        \n",
    "    codon_change = get_codon(row[\"pos\"], ref_name, row[\"alt\"])\n",
    "    glyco_snps.loc[i, \"ref_aa\"] = str(codon_change[0])\n",
    "    glyco_snps.loc[i, \"alt_aa\"] = str(codon_change[1])\n",
    "    \n",
    "glyco_snps[\"synonymous\"] = (glyco_snps[\"ref_aa\"] == glyco_snps[\"alt_aa\"]).astype(int)\n",
    "\n",
    "# add the amino acid position (1-indexed)\n",
    "glyco_snps[\"aa_pos\"] = [get_aa_num(pos) for pos in glyco_snps[\"pos\"]]\n",
    "\n",
    "# glyco_snps.to_csv(\"snp_calling/glyco.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5b956871-4c5a-4c89-aa11-c38feabc39de",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, row in phospho_snps.iterrows():\n",
    "    \n",
    "    if row[\"aln\"] == \"BGD\":\n",
    "        ref_name = phospho_bgd_ref\n",
    "    else:\n",
    "        ref_name = phospho_mys_ref\n",
    "        \n",
    "    codon_change = get_codon(row[\"pos\"], ref_name, row[\"alt\"])\n",
    "    phospho_snps.loc[i, \"ref_aa\"] = str(codon_change[0])\n",
    "    phospho_snps.loc[i, \"alt_aa\"] = str(codon_change[1])\n",
    "    \n",
    "phospho_snps[\"synonymous\"] = (phospho_snps[\"ref_aa\"] == phospho_snps[\"alt_aa\"]).astype(int)\n",
    "\n",
    "# add the amino acid position (1-indexed)\n",
    "phospho_snps[\"aa_pos\"] = [get_aa_num(pos) for pos in phospho_snps[\"pos\"]]\n",
    "\n",
    "#phospho_snps.to_csv(\"snp_calling/phospho.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ec15d9e-8752-4abf-baa6-53b9136011be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# vcf_reader = vcf.Reader(open(\"alignments/G_BGD_SNPs.vcf\", \"r\"))\n",
    "\n",
    "# for record in vcf_reader:\n",
    "    \n",
    "#     # convert form VCF objects to strings\n",
    "#     alt_alleles = [str(char) for char in record.ALT]\n",
    "    \n",
    "#     if \"<*>\" in alt_alleles:\n",
    "#         alt_alleles.remove(\"<*>\")\n",
    "#         print(alt_alleles)\n",
    "#     # else:\n",
    "#     #     print(alt_alleles)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "529b413e-da6b-4116-ad0f-8163a972a871",
   "metadata": {},
   "source": [
    "# Further investigation of SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "bd5a91d3-3994-4c94-a74e-e00f395627f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "glyco_snps = pd.read_csv(\"snp_calling/glyco.csv\")\n",
    "phospho_snps = pd.read_csv(\"snp_calling/phospho.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a2f72a28-0fa9-4944-b977-0541d159e313",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>aln</th>\n",
       "      <th>ref_aa</th>\n",
       "      <th>alt_aa</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>aa_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>BGD</td>\n",
       "      <td>T</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>199</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>BGD</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>0</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>246</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BGD</td>\n",
       "      <td>M</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>625</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BGD</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>682</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>BGD</td>\n",
       "      <td>Y</td>\n",
       "      <td>H</td>\n",
       "      <td>0</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>734</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>BGD</td>\n",
       "      <td>S</td>\n",
       "      <td>F</td>\n",
       "      <td>0</td>\n",
       "      <td>245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>889</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BGD</td>\n",
       "      <td>V</td>\n",
       "      <td>M</td>\n",
       "      <td>0</td>\n",
       "      <td>297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>910</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>BGD</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>0</td>\n",
       "      <td>304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>1127</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>BGD</td>\n",
       "      <td>K</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>1142</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>BGD</td>\n",
       "      <td>N</td>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>1148</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>BGD</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>0</td>\n",
       "      <td>383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>1172</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>BGD</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>0</td>\n",
       "      <td>391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>1211</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>BGD</td>\n",
       "      <td>N</td>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>1412</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>BGD</td>\n",
       "      <td>T</td>\n",
       "      <td>I</td>\n",
       "      <td>0</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>1432</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>BGD</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>0</td>\n",
       "      <td>478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>1706</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>BGD</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>0</td>\n",
       "      <td>569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>556</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>MYS</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>814</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>MYS</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>1330</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>MYS</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>0</td>\n",
       "      <td>444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     pos ref alt  aln ref_aa alt_aa  synonymous  aa_pos\n",
       "0      8   C   T  BGD      T      I           0       3\n",
       "9    199   A   G  BGD      I      V           0      67\n",
       "11   246   G   A  BGD      M      I           0      82\n",
       "29   625   G   A  BGD      V      I           0     209\n",
       "31   682   T   C  BGD      Y      H           0     228\n",
       "32   734   C   T  BGD      S      F           0     245\n",
       "35   889   G   A  BGD      V      M           0     297\n",
       "36   910   A   G  BGD      I      V           0     304\n",
       "44  1127   A   C  BGD      K      T           0     376\n",
       "46  1142   A   G  BGD      N      S           0     381\n",
       "47  1148   C   T  BGD      P      L           0     383\n",
       "49  1172   A   G  BGD      K      R           0     391\n",
       "51  1211   A   G  BGD      N      S           0     404\n",
       "58  1412   C   T  BGD      T      I           0     471\n",
       "59  1432   A   G  BGD      N      D           0     478\n",
       "67  1706   A   G  BGD      K      R           0     569\n",
       "75   556   A   G  MYS      N      D           0     186\n",
       "77   814   A   G  MYS      T      A           0     272\n",
       "78  1330   A   G  MYS      I      V           0     444"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glyco_snps.query(\"synonymous==0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b30e4469-43b5-443f-be9a-2632b91fd4bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping to use for one-hot encoding. No nucleotide = missing because already checked that there are no indels in these seque\n",
    "BASE_TO_COLUMN = {'A': 0, 'C': 1, 'T': 2, 'G': 3, 'N': 4}\n",
    "COLUMN_TO_BASE = {0: \"A\", 1: \"C\", 2: \"T\", 3: \"G\"}\n",
    "\n",
    "# Get one hot vector\n",
    "def get_one_hot(sequence):\n",
    "\n",
    "    if not pd.isnull(sequence):\n",
    "        seq_in_index = np.array([BASE_TO_COLUMN.get(b, b) for b in sequence])\n",
    "\n",
    "        one_hot = np.zeros((len(sequence), 4))\n",
    "\n",
    "        # only get the non gap sites because otherwise we get indexing errors\n",
    "        ind_non_gap = np.where(seq_in_index != 4)[0]\n",
    "\n",
    "        # Assign the found positions to 1\n",
    "        one_hot[ind_non_gap, seq_in_index[ind_non_gap]] = 1\n",
    "\n",
    "        return one_hot\n",
    "    else:\n",
    "        return np.nan\n",
    "\n",
    "# metadata file previously generated for all sequences\n",
    "metadata = pd.read_csv(\"metadata_all.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
