{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b1da87d2-25a3-41ee-984b-e32be6043c1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as st\n",
    "import vcf\n",
    "import vcf.filters\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import gzip\n",
    "import shutil\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82d8b21e-ff22-4eda-9975-dc54a96e90c9",
   "metadata": {},
   "source": [
    "# Extracting and analyzing data from the VCF files\n",
    "\n",
    "## Don't need to do any QC filtering because that's for reads from FASTQ files\n",
    "\n",
    "## These are all post-QC sequences in FASTA files\n",
    "\n",
    "```bash\n",
    "bash snp_calling/call_snps.sh sequences/genome/AY988601.1_BGD_CDS.fasta seq_for_analysis/P_whitmer_BGD.fasta alignments P_whitmer_BGD\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0f7ed397-b340-4a8a-923c-1004df49954c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_vcf_snps(fName):\n",
    "    \n",
    "    if \"G_\" in fName:\n",
    "        chrom = \"gene=G\"\n",
    "        protein = \"G\"\n",
    "    elif \"P_\" in fName:\n",
    "        chrom = \"gene=P\"\n",
    "        protein = \"P\"\n",
    "    \n",
    "    pos = []\n",
    "    ref = []\n",
    "    alt = []    \n",
    "    \n",
    "    vcf_reader = vcf.Reader(open(fName, 'r'))\n",
    "\n",
    "    for record in vcf_reader:\n",
    "        \n",
    "        # check that all variants are coming from the correct gene. Might be overkill, but just another sanity check\n",
    "        if chrom not in record.CHROM or record.CHROM != chrom:\n",
    "            raise ValueError(f\"SNP found outside of {chrom}\")\n",
    "            \n",
    "        # convert form VCF objects to strings to do string comparison\n",
    "        alt_alleles = [str(char) for char in record.ALT]\n",
    "\n",
    "        if \"<*>\" in alt_alleles:\n",
    "            alt_alleles.remove(\"<*>\")\n",
    "        alt_alleles = \"\".join(alt_alleles)\n",
    "            \n",
    "        pos.append(record.POS)\n",
    "        ref.append(record.REF)\n",
    "        alt.append(alt_alleles)\n",
    "        \n",
    "    return pd.DataFrame({\"pos\": pos, \"ref\": ref, \"alt\": alt})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5b65cdf0-2734-460b-86db-6bc50f132d8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_bgd = parse_vcf_snps(\"alignments/P_whitmer_BGD_SNPS.vcf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "147fb336-4c22-47eb-9407-b6da0f8a0e9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>99</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>167</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>175</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>181</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>190</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>2094</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>2097</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>2101</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>2112</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>2121</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     pos ref alt\n",
       "0     99   T   C\n",
       "1    167   C   T\n",
       "2    175   G   A\n",
       "3    181   G   A\n",
       "4    190   T   C\n",
       "..   ...  ..  ..\n",
       "83  2094   A   T\n",
       "84  2097   T   C\n",
       "85  2101   G   A\n",
       "86  2112   C   T\n",
       "87  2121   T   C\n",
       "\n",
       "[88 rows x 3 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_bgd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c6928bd2-3708-41cb-ad9a-59172884b709",
   "metadata": {},
   "outputs": [],
   "source": [
    "P_whitmer = [(seq.id, str(seq.seq)) for seq in SeqIO.parse(\"seq_for_analysis/P_whitmer_BGD.fasta\", \"fasta\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f52f4d5-454a-4fc1-8f1b-d1c9049497c5",
   "metadata": {},
   "source": [
    "# Next, encode every sequence in `P_whitmer` as a binary variable representing reference (0) or alternative (1) allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bfe8c25-4623-4feb-a217-a3d29f487a28",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b30e4469-43b5-443f-be9a-2632b91fd4bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapping to use for one-hot encoding. No nucleotide = missing because already checked that there are no indels in these seque\n",
    "BASE_TO_COLUMN = {'A': 0, 'C': 1, 'T': 2, 'G': 3, 'N': 4}\n",
    "COLUMN_TO_BASE = {0: \"A\", 1: \"C\", 2: \"T\", 3: \"G\"}\n",
    "\n",
    "# Get one hot vector\n",
    "def get_one_hot(sequence):\n",
    "\n",
    "    if not pd.isnull(sequence):\n",
    "        seq_in_index = np.array([BASE_TO_COLUMN.get(b, b) for b in sequence])\n",
    "\n",
    "        one_hot = np.zeros((len(sequence), 4))\n",
    "\n",
    "        # only get the non gap sites because otherwise we get indexing errors\n",
    "        ind_non_gap = np.where(seq_in_index != 4)[0]\n",
    "\n",
    "        # Assign the found positions to 1\n",
    "        one_hot[ind_non_gap, seq_in_index[ind_non_gap]] = 1\n",
    "\n",
    "        return one_hot\n",
    "    else:\n",
    "        return np.nan\n",
    "\n",
    "# metadata file previously generated for all sequences\n",
    "metadata = pd.read_csv(\"metadata_all.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
