{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1c73ebe8-cc31-49ff-a082-924f8cbbf11f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2c9479d1-ead5-48b5-bedb-aedb67d94d3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Malaysian strain genome size: 18246\n"
     ]
    }
   ],
   "source": [
    "#for seq_record in SeqIO.parse(\"sequences/genome/AY029768.1_MYS.fasta\", \"fasta\"):\n",
    "for seq_record in SeqIO.parse(\"sequences/genome/NC_00278.1_MYS.fasta\", \"fasta\"):\n",
    "    \n",
    "    # just wanted to do a sanity check\n",
    "    print(f\"Malaysian strain genome size: {len(seq_record)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ae679963-189d-47a6-87e8-40324dbfe8a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bangaldeshi strain genome size: 18252\n"
     ]
    }
   ],
   "source": [
    "for seq_record in SeqIO.parse(\"sequences/genome/AY988601.1_BGD.fasta\", \"fasta\"):\n",
    "    \n",
    "    # just wanted to do a sanity check\n",
    "    print(f\"Bangaldeshi strain genome size: {len(seq_record)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c76692f9-7cdd-4d51-8687-0b9027609539",
   "metadata": {},
   "source": [
    "# 1. BLAST the glycoprotein and phosphoprotein sequences to get all sequences from NCBI\n",
    "\n",
    "Used reference sequences `sequences/PG/ref_G_seq` and `sequences/PG/ref_P_seq` for the glycoprotein and phosphoproteins. Queried these on BLAST using <i>Henipaviruses</i> as the search organism. \n",
    "\n",
    "83 full search results for G (could be individual protein, multiple, or full genome) stored at `sequences/PG/glyco_blast_83.fasta`, and 80 full search results for P stored at `sequences/PG/phospho_blast_80.fasta`. Removed duplicates in Geneious.\n",
    "\n",
    "Descriptions stored at `sequences/glyco_blast_descriptions.csv` and `sequences/phospho_blast_descriptions.csv`. Need this to get the isolate locations and other metadata."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c678a03e-520d-4bc6-9aa4-904f9aef62bf",
   "metadata": {},
   "source": [
    "```bash\n",
    "bowtie2-build sequences/genome/AY029768.1_MYS_CDS.fna MYS_cds\n",
    "bowtie2 -x MYS_cds -f sequences/PG/G_seqs.fasta -S G_aln.sam\n",
    "```\n",
    "\n",
    "From the above, found that two glycoprotein sequences did not align to either genome because they have too many ambiguous nucleotides in them. These are <b>MN549407</b> and <b>MN549410</b>. They were removed from the .fasta files.\n",
    "\n",
    "The sequence headers were updated, and the V, W, and C proteins were removed because they are duplicates of the phosphoprotein sequence. <b>This was found to cause problems when aligning the glycoprotein sequences</b>. It makes sense that you shouldn't have redundant sequences in a reference sequence, but I'm not sure why it didn't cause problems for the phosphoprotein alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5e78b1ff-bd8c-4c45-9987-c705f2dc8c96",
   "metadata": {},
   "outputs": [],
   "source": [
    "glyco_descript = pd.read_csv(\"sequences/glyco_blast_descriptions.csv\")\n",
    "phospho_descript = pd.read_csv(\"sequences/phospho_blast_descriptions.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "91043c73-fc2b-47de-ae36-d4fc64be9cce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Scientific Name</th>\n",
       "      <th>Max Score</th>\n",
       "      <th>Total Score</th>\n",
       "      <th>Query Cover</th>\n",
       "      <th>E value</th>\n",
       "      <th>Per. ident</th>\n",
       "      <th>Acc. Len</th>\n",
       "      <th>Accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Nipah virus N gene, P gene, M gene, F gene, G ...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>3341</td>\n",
       "      <td>3341</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>14973</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Nipah virus strain Malaysian glycoprotein (G) ...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>3341</td>\n",
       "      <td>3341</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1809</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Nipah henipavirus isolate 808651, complete genome</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>3275</td>\n",
       "      <td>3275</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.34</td>\n",
       "      <td>18231</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Nipah virus isolate IRF0158, partial genome</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>3269</td>\n",
       "      <td>3269</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.28</td>\n",
       "      <td>18214</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Nipah virus isolate IRF0160, partial genome</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>3269</td>\n",
       "      <td>3269</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.28</td>\n",
       "      <td>18212</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Nipah henipavirus isolate MCL-19-BAT-572-9, pa...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>1428</td>\n",
       "      <td>2341</td>\n",
       "      <td>87%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>93.29</td>\n",
       "      <td>18132</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Nipah virus glycoprotein (G) gene, partial cds</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>1363</td>\n",
       "      <td>1363</td>\n",
       "      <td>52%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.53</td>\n",
       "      <td>954</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Nipah henipavirus isolate MCL-19-BAT-574-7, pa...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>1214</td>\n",
       "      <td>2088</td>\n",
       "      <td>79%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.83</td>\n",
       "      <td>17815</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Nipah henipavirus isolate MCL-19-BAT-574-5, pa...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>1199</td>\n",
       "      <td>2433</td>\n",
       "      <td>91%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>93.05</td>\n",
       "      <td>18131</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>Nipah henipavirus isolate MCL-19-BAT-572-6, pa...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>1195</td>\n",
       "      <td>1849</td>\n",
       "      <td>70%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.56</td>\n",
       "      <td>17813</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Description    Scientific Name  \\\n",
       "0   Nipah virus N gene, P gene, M gene, F gene, G ...  Nipah henipavirus   \n",
       "1   Nipah virus strain Malaysian glycoprotein (G) ...  Nipah henipavirus   \n",
       "2   Nipah henipavirus isolate 808651, complete genome  Nipah henipavirus   \n",
       "3         Nipah virus isolate IRF0158, partial genome  Nipah henipavirus   \n",
       "4         Nipah virus isolate IRF0160, partial genome  Nipah henipavirus   \n",
       "..                                                ...                ...   \n",
       "78  Nipah henipavirus isolate MCL-19-BAT-572-9, pa...  Nipah henipavirus   \n",
       "79     Nipah virus glycoprotein (G) gene, partial cds  Nipah henipavirus   \n",
       "80  Nipah henipavirus isolate MCL-19-BAT-574-7, pa...  Nipah henipavirus   \n",
       "81  Nipah henipavirus isolate MCL-19-BAT-574-5, pa...  Nipah henipavirus   \n",
       "82  Nipah henipavirus isolate MCL-19-BAT-572-6, pa...  Nipah henipavirus   \n",
       "\n",
       "    Max Score  Total Score Query Cover  E value  Per. ident  Acc. Len  \\\n",
       "0        3341         3341        100%      0.0      100.00     14973   \n",
       "1        3341         3341        100%      0.0      100.00      1809   \n",
       "2        3275         3275        100%      0.0       99.34     18231   \n",
       "3        3269         3269        100%      0.0       99.28     18214   \n",
       "4        3269         3269        100%      0.0       99.28     18212   \n",
       "..        ...          ...         ...      ...         ...       ...   \n",
       "78       1428         2341         87%      0.0       93.29     18132   \n",
       "79       1363         1363         52%      0.0       92.53       954   \n",
       "80       1214         2088         79%      0.0       92.83     17815   \n",
       "81       1199         2433         91%      0.0       93.05     18131   \n",
       "82       1195         1849         70%      0.0       92.56     17813   \n",
       "\n",
       "                                          Accession    \n",
       "0   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "1   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "2   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "3   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "4   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "..                                                ...  \n",
       "78  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "79  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "80  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "81  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "82  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "\n",
       "[83 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glyco_descript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "67c3940f-0d82-44b1-b5c7-62aa312d38eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Scientific Name</th>\n",
       "      <th>Max Score</th>\n",
       "      <th>Total Score</th>\n",
       "      <th>Query Cover</th>\n",
       "      <th>E value</th>\n",
       "      <th>Per. ident</th>\n",
       "      <th>Acc. Len</th>\n",
       "      <th>Accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Nipah virus N gene, P gene, M gene, F gene, G ...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>3341</td>\n",
       "      <td>3341</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>14973</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Nipah virus strain Malaysian glycoprotein (G) ...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>3341</td>\n",
       "      <td>3341</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1809</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Nipah henipavirus isolate 808651, complete genome</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>3275</td>\n",
       "      <td>3275</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.34</td>\n",
       "      <td>18231</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Nipah virus isolate IRF0158, partial genome</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>3269</td>\n",
       "      <td>3269</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.28</td>\n",
       "      <td>18214</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Nipah virus isolate IRF0160, partial genome</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>3269</td>\n",
       "      <td>3269</td>\n",
       "      <td>100%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.28</td>\n",
       "      <td>18212</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Nipah henipavirus isolate MCL-19-BAT-572-9, pa...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>1428</td>\n",
       "      <td>2341</td>\n",
       "      <td>87%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>93.29</td>\n",
       "      <td>18132</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Nipah virus glycoprotein (G) gene, partial cds</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>1363</td>\n",
       "      <td>1363</td>\n",
       "      <td>52%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.53</td>\n",
       "      <td>954</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Nipah henipavirus isolate MCL-19-BAT-574-7, pa...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>1214</td>\n",
       "      <td>2088</td>\n",
       "      <td>79%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.83</td>\n",
       "      <td>17815</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Nipah henipavirus isolate MCL-19-BAT-574-5, pa...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>1199</td>\n",
       "      <td>2433</td>\n",
       "      <td>91%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>93.05</td>\n",
       "      <td>18131</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>Nipah henipavirus isolate MCL-19-BAT-572-6, pa...</td>\n",
       "      <td>Nipah henipavirus</td>\n",
       "      <td>1195</td>\n",
       "      <td>1849</td>\n",
       "      <td>70%</td>\n",
       "      <td>0.0</td>\n",
       "      <td>92.56</td>\n",
       "      <td>17813</td>\n",
       "      <td>=HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Description    Scientific Name  \\\n",
       "0   Nipah virus N gene, P gene, M gene, F gene, G ...  Nipah henipavirus   \n",
       "1   Nipah virus strain Malaysian glycoprotein (G) ...  Nipah henipavirus   \n",
       "2   Nipah henipavirus isolate 808651, complete genome  Nipah henipavirus   \n",
       "3         Nipah virus isolate IRF0158, partial genome  Nipah henipavirus   \n",
       "4         Nipah virus isolate IRF0160, partial genome  Nipah henipavirus   \n",
       "..                                                ...                ...   \n",
       "78  Nipah henipavirus isolate MCL-19-BAT-572-9, pa...  Nipah henipavirus   \n",
       "79     Nipah virus glycoprotein (G) gene, partial cds  Nipah henipavirus   \n",
       "80  Nipah henipavirus isolate MCL-19-BAT-574-7, pa...  Nipah henipavirus   \n",
       "81  Nipah henipavirus isolate MCL-19-BAT-574-5, pa...  Nipah henipavirus   \n",
       "82  Nipah henipavirus isolate MCL-19-BAT-572-6, pa...  Nipah henipavirus   \n",
       "\n",
       "    Max Score  Total Score Query Cover  E value  Per. ident  Acc. Len  \\\n",
       "0        3341         3341        100%      0.0      100.00     14973   \n",
       "1        3341         3341        100%      0.0      100.00      1809   \n",
       "2        3275         3275        100%      0.0       99.34     18231   \n",
       "3        3269         3269        100%      0.0       99.28     18214   \n",
       "4        3269         3269        100%      0.0       99.28     18212   \n",
       "..        ...          ...         ...      ...         ...       ...   \n",
       "78       1428         2341         87%      0.0       93.29     18132   \n",
       "79       1363         1363         52%      0.0       92.53       954   \n",
       "80       1214         2088         79%      0.0       92.83     17815   \n",
       "81       1199         2433         91%      0.0       93.05     18131   \n",
       "82       1195         1849         70%      0.0       92.56     17813   \n",
       "\n",
       "                                          Accession    \n",
       "0   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "1   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "2   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "3   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "4   =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "..                                                ...  \n",
       "78  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "79  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "80  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "81  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "82  =HYPERLINK(\"https://www.ncbi.nlm.nih.gov/nucle...  \n",
       "\n",
       "[83 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glyco_descript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "42d9c1a0-d1db-40a5-aca1-c24e53761367",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_fasta(fName):\n",
    "    \n",
    "    lengths = []\n",
    "    names = []\n",
    "    seqs = []\n",
    "\n",
    "    for seq_record in SeqIO.parse(fName, \"fasta\"):\n",
    "\n",
    "        if \"_\" in seq_record.id:\n",
    "            names.append(seq_record.id.split(\"_\")[0])\n",
    "        else:\n",
    "            names.append(seq_record.id)\n",
    "        seqs.append(seq_record.seq)\n",
    "        lengths.append(len(seq_record))\n",
    "\n",
    "    print(f\"Found {len(seqs)} sequences in {os.path.basename(fName)}\")\n",
    "    \n",
    "    res_dict = dict(zip(np.array(names), seqs))\n",
    "    \n",
    "    return dict(sorted(res_dict.items(), key=lambda x: x[0])), np.array(lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8ca1a1d3-b887-40b9-b3cb-adda5e0c414e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 79 sequences in G_seqs.fasta\n",
      "Found 79 sequences in P_seqs.fasta\n"
     ]
    }
   ],
   "source": [
    "G = read_fasta(\"sequences/PG/G_seqs.fasta\")\n",
    "P = read_fasta(\"sequences/PG/P_seqs.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5468beb2-d481-498f-8bec-044c02cfa5ad",
   "metadata": {},
   "source": [
    "# 2. Single nucleotide polymorphism (SNP) calling\n",
    "\n",
    "## End goal is to convert the FASTA file into a VCF (variant calling format) file\n",
    "\n",
    "<b>Amazing tutorial:</b> https://www.ebi.ac.uk/sites/ebi.ac.uk/files/content.ebi.ac.uk/materials/2014/140217_AgriOmics/dan_bolser_snp_calling.pdf\n",
    "\n",
    "Adapted the following steps from it"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73f8acc6-9608-4bc2-b7f6-7894c41fa16b",
   "metadata": {},
   "source": [
    "<b></b>\n",
    "```bash\n",
    "bwa index sequences/genome/AY029768.1_MYS_CDS.fna\n",
    "```\n",
    "\n",
    "<b>Perform the alignment</b>\n",
    "```bash\n",
    "bwa aln sequences/genome/AY029768.1_MYS_CDS.fna sequences/PG/P_seqs.fasta > alignments/P_MYS_aln.sai\n",
    "```\n",
    "\n",
    "<b>Convert to SAM file format, which is human-readable</b>\n",
    "```bash\n",
    "bwa samse sequences/genome/AY029768.1_MYS_CDS.fna alignments/P_MYS_aln.sai sequences/PG/P_seqs.fasta > alignments/P_MYS_aln.sam\n",
    "```\n",
    "\n",
    "<b>Convert SAM to BAM and sort the BAM file</b>\n",
    "```bash\n",
    "   samtools view -b alignments/G_BGD_aln.sam > alignments/G_BGD_aln.bam\n",
    "   samtools sort alignments/G_BGD_aln.bam -o alignments/G_BGD_aln_sorted.bam\n",
    "```\n",
    "\n",
    "<b>Index the genome file again with `samtools`</b>\n",
    "```bash\n",
    "samtools faidx sequences/genome/AY029768.1_MYS_CDS.fna\n",
    "```\n",
    "\n",
    "<b>Run 'mpileup' to generate VCF format</b>\n",
    "```bash\n",
    "bcftools mpileup -f sequences/genome/AY988601.1_BGD_CDS.fna alignments/P_MYS_aln_sorted.bam > alignments/P_MYS_aln.bcf\n",
    "bcftools mpileup -f sequences/genome/AY988601.1_BGD_CDS.fna alignments/G_BGD_aln_sorted.bam > alignments/G_BGD_aln.bcf\n",
    "```\n",
    "\n",
    "<b>Call SNPs</b>\n",
    "```bash\n",
    "bcftools view -v snps alignments/P_MYS_aln.bcf > alignments/P_MYS_SNPs.vcf\n",
    "bcftools view -v snps alignments/G_MYS_aln.bcf > alignments/G_MYS_SNPs.vcf\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe7a4ce1-35f3-4af1-9a29-68d91cf2d986",
   "metadata": {},
   "source": [
    "## Exact code to run:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85cfb934-175b-4fca-b3d1-45262ceb4610",
   "metadata": {},
   "source": [
    "```\n",
    "bash align_make_vcf.sh \"sequences/genome/AY988601.1_BGD_CDS.fna\" \"sequences/PG/P_seqs.fasta\" \"alignments\" \"P_BGD\"\n",
    "bash align_make_vcf.sh \"sequences/genome/NC_00278.1_MYS_CDS.fna\" \"sequences/PG/P_seqs.fasta\" \"alignments\" \"P_MYS\"\n",
    "\n",
    "bash align_make_vcf.sh \"sequences/genome/AY988601.1_BGD_CDS.fna\" \"sequences/PG/G_seqs.fasta\" \"alignments\" \"G_BGD\"\n",
    "bash align_make_vcf.sh \"sequences/genome/NC_00278.1_MYS_CDS.fna\" \"sequences/PG/G_seqs.fasta\" \"alignments\" \"G_MYS\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c00c4319-ca52-4b78-abfc-ce47735179d5",
   "metadata": {},
   "source": [
    "# 3. Make trees using the G and P CDS's\n",
    "\n",
    "```bash\n",
    "fasttree -nt sequences/PG/G_seqs.fasta > trees/G_cds.nwk\n",
    "fasttree -nt sequences/PG/P_seqs.fasta > trees/P_cds.nwk\n",
    "```\n",
    "\n",
    "Another reference: https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionVI/lessons/01_alignment.html\n",
    "\n",
    "# 4. Deduplicate and remake trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "44e0299b-e513-44d3-9c1c-8704547e5fbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "G_seqs = [(seq_record.id, str(seq_record.seq)) for seq_record in SeqIO.parse(\"sequences/PG/G_seqs.fasta\", \"fasta\")]\n",
    "P_seqs = [(seq_record.id, str(seq_record.seq)) for seq_record in SeqIO.parse(\"sequences/PG/P_seqs.fasta\", \"fasta\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "772e7cdc-e8d8-459d-b06d-97b9c5af1fae",
   "metadata": {},
   "source": [
    "## More sanity checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "22bf0faf-ebfd-4439-b2e9-bb169429d699",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "76\n",
      "{'AF238467', 'HM545086', 'AY858111'}\n",
      "{'MN549407', 'MN549410', 'HM545087'}\n"
     ]
    }
   ],
   "source": [
    "G_ids = np.array(list(zip(*G_seqs))[0])\n",
    "P_ids = np.array(list(zip(*P_seqs))[0])\n",
    "\n",
    "# 76 / 79 sequences are the same. 3 are not\n",
    "print(len(set(G_ids).intersection(P_ids)))\n",
    "\n",
    "# Sequences with G only\n",
    "print(set(G_ids) - set(P_ids))\n",
    "\n",
    "# HM545087 is P only. MN549407 and MN549410 were the very low quality G sequences that were removed\n",
    "# according to the entries, they are partial genomes isolated from bats, probably explains the low quality\n",
    "print(set(P_ids) - set(G_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "76d88884-0ee6-4749-881d-84e727f984ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46, 50)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs_G_df = pd.DataFrame(list(zip(*G_seqs))[-1]).rename(columns={0:\"Seq\"})\n",
    "seqs_G_df[\"ID\"] = G_ids\n",
    "\n",
    "seqs_P_df = pd.DataFrame(list(zip(*P_seqs))[-1]).rename(columns={0:\"Seq\"})\n",
    "seqs_P_df[\"ID\"] = P_ids\n",
    "\n",
    "# keeps only the first occurrence. So there are 46 unique glycoprotein sequences and 50 unique phosphoprotein sequences\n",
    "len(seqs_G_df[\"Seq\"].drop_duplicates()), len(seqs_P_df[\"Seq\"].drop_duplicates())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44e7edab-d6fd-4be4-88f9-2e397dd987d2",
   "metadata": {},
   "source": [
    "## The next cell writes the unique sequences to new Fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "84d0f5c1-2114-4fd1-9e42-8b07722ed585",
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep_G = seqs_G_df.drop_duplicates(\"Seq\")\n",
    "# keep_P = seqs_P_df.drop_duplicates(\"Seq\")\n",
    "\n",
    "# with open(\"sequences/PG/G_deduplicated.fasta\", \"w+\") as file:\n",
    "    \n",
    "#     for _, row in keep_G.iterrows():\n",
    "#         file.write(\">\" + row[\"ID\"] + \"\\n\")\n",
    "#         file.write(row[\"Seq\"] + \"\\n\")\n",
    "        \n",
    "# with open(\"sequences/PG/P_deduplicated.fasta\", \"w+\") as file:\n",
    "    \n",
    "#     for _, row in keep_P.iterrows():\n",
    "#         file.write(\">\" + row[\"ID\"] + \"\\n\")\n",
    "#         file.write(row[\"Seq\"] + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "16c9b082-520f-443d-947c-8e9daf464cbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "G_dedup = [(seq_record.id, str(seq_record.seq)) for seq_record in SeqIO.parse(\"sequences/PG/G_deduplicated.fasta\", \"fasta\")]\n",
    "P_dedup = [(seq_record.id, str(seq_record.seq)) for seq_record in SeqIO.parse(\"sequences/PG/P_deduplicated.fasta\", \"fasta\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4317de71-ffa9-429e-a27c-212c24a162b8",
   "metadata": {},
   "source": [
    "```bash\n",
    "fasttree -nt sequences/PG/G_deduplicated.fasta > trees/G_dedup.nwk\n",
    "fasttree -nt sequences/PG/P_deduplicated.fasta > trees/P_dedup.nwk\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d29e2bbb-1ccf-4282-80d0-9d03cefcde09",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
